spec/daffy_lib/caching_encryptor_spec.rb

```diff
@@ -0,0 +1,238 @@
+    end
+
+    it 'raises an EncryptionFailedException on RbNaCl CryptoError' do
+      allow(PorkyLib::Symmetric.instance).to receive(:encrypt_with_key).and_raise(RbNaCl::CryptoError)
+
+      expect { caching_encryptor.zt_encrypt(*encryption_args) }.to raise_exception(DaffyLib::CachingEncryptor::EncryptionFailedException)
+    end
+
+    it 'raises an EncryptionFailedException on RbNaCl LengthError' do
+      allow(PorkyLib::Symmetric.instance).to receive(:encrypt_with_key).and_raise(RbNaCl::LengthError)
+
+      expect { caching_encryptor.zt_encrypt(*encryption_args) }.to raise_exception(DaffyLib::CachingEncryptor::EncryptionFailedException)
+    end
+
+    it 'returns a correct valid JSON struct' do
+      encryption_info = caching_encryptor.zt_encrypt(*encryption_args)
+
+      expect { JSON.parse(encryption_info) }.not_to raise_exception
+    end
+
+    it 'returns a correct JSON struct with the correct fields' do
+      encryption_info = JSON.parse(caching_encryptor.zt_encrypt(*encryption_args), symbolize_names: true)
+
+      expect(encryption_info).to have_key(:key_guid)
+      expect(encryption_info).to have_key(:key)
+      expect(encryption_info).to have_key(:data)
+      expect(encryption_info).to have_key(:nonce)
+    end
+
+    it 'returns a correct JSON struct with the correct values' do
+      encryption_info = JSON.parse(caching_encryptor.zt_encrypt(*encryption_args), symbolize_names: true)
+
+      expect(encryption_info[:key_guid]).to eq(encryption_key.guid)
+      expect(encryption_info[:key]).to eq(encryption_key.encrypted_data_encryption_key)
+      expect(encryption_info[:data]).to eq(Base64.encode64(ciphertext))
+      expect(encryption_info[:nonce]).to eq(Base64.encode64(nonce))
+    end
+  end
+
+  describe '#decrypt' do
+    decryption_params.each do |param|
+      it "raises an InvalidParameterException when #{param} is nil" do
+        expect do
+          caching_encryptor.zt_decrypt(*generate_invalid_decrypt_params(param, nil))
+        end.to raise_exception(DaffyLib::CachingEncryptor::InvalidParameterException)
+      end
+
+      it "raises an InvalidParameterException when #{param} is empty" do
+        expect do
+          caching_encryptor.zt_decrypt(*generate_invalid_decrypt_params(param, ''))
+        end.to raise_exception(DaffyLib::CachingEncryptor::InvalidParameterException)
+      end
+    end
+
+    it 'raises a DecryptionFailedException on a JSON parse error' do
+      decryption_args = [{
+        value: 'this is not { valid => json }: here',
+        expires_in: expires_in,
+        cmk_key_id: cmk_key_id
+      }]
+
+      expect { caching_encryptor.zt_decrypt(*decryption_args) }.to raise_exception(DaffyLib::CachingEncryptor::DecryptionFailedException)
+    end
+
+    # rubocop:disable ExampleLength
+    it 'raises a DecryptionFailedException on an EncryptionKey not found error' do
```

Comment:
This is one of the cases where we should fallback to use the key in the json blob

Reply:
Noted.